### PR TITLE
Refactor module version/GUID comparison logic

### LINKED ISSUE #7496: Import-Module on path ignores version constraints
Steps to reproduce
------------------

**Note**: The repro uses a relative path, but an absolute path works too; just insert `$PWD` instead of the `.` below.

```powershell
> mkdir ./AModule
> New-ModuleManifest -Path ./AModule/AModule.psd1 -ModuleVersion '2.0'
> Import-Module -Path ./AModule/
> Import-Module -PassThru -FullQualifiedName @{ ModuleName = "./AModule/AModule.psd1"; ModuleVersion = '3.0' } # Also try the absolute path - same outcome
```

Expected behavior
-----------------

```none
Import-Module : The specified module './AModule/AModule.psd1' with version '3.0' was not loaded because no valid module file was found in any module directory.
At line:1 char:1
+ Import-Module -PassThru -FullyQualifiedName @{ ModuleName = './AModul ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ResourceUnavailable: (./AModule:String) [Import-Module], FileNotFoundException
+ FullyQualifiedErrorId : Modules_ModuleWithVersionNotFound,Microsoft.PowerShell.Commands.ImportModuleCommand

```

Actual behavior
---------------

```none

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Manifest   2.0        AModule                             

```

Environment data
----------------

```powershell
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.1.0-preview.4
PSEdition                      Core
GitCommitId                    6.1.0-preview.4-61-g1243891ea8c04609ce21d0b31d40945fd46a3576
OS                             Linux 4.15.0-30-generic #32-Ubuntu SMP Thu Jul 26 17:42:43 UTC 2018
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0


```

## More information

**EDIT**: Just realised this is for re-imports only (which is good!)

The important note is that importing the same module off the module path will fail:

```powershell
> $env:PSModulePath += [System.IO.Path]::PathSeparator + $PWD
> Import-Module -PassThru -FullyQualifiedName @{ ModuleName = 'AModule'; ModuleVersion = '3.0' }
Import-Module : The specified module 'AModule' with version '3.0' was not loaded because no valid module file was found in any module directory.
At line:1 char:1
+ Import-Module -PassThru -FullyQualifiedName @{ ModuleName = 'AModule' ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ResourceUnavailable: (AModule:String) [Import-Module], FileNotFoundException
+ FullyQualifiedErrorId : Modules_ModuleWithVersionNotFound,Microsoft.PowerShell.Commands.ImportModuleCommand
 
```

This behaviour is caused by this logic:
https://github.com/PowerShell/PowerShell/blob/1243891ea8c04609ce21d0b31d40945fd46a3576/src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs#L589-L595

Note that [it's impossible to instantiate a ModuleSpecification with all checked fields not `null`](https://github.com/PowerShell/PowerShell/blob/1243891ea8c04609ce21d0b31d40945fd46a3576/src/System.Management.Automation/engine/Modules/ModuleSpecification.cs#L155-L165), so this check always succeeds.

This is addressed by https://github.com/PowerShell/PowerShell/pull/7125, but this discovery means it's a breaking change.

### PR DESCRIPTION
**Pull request by rjmholt**
## PR Summary

**TLDR**: Factors out module GUID/version checking logic to all use the same code path.

This PR routes all of the module GUID/version checking through `ModuleIntrinsics.IsModuleMatchingSpec()` (or a sub-method of it), so that all the version checking is consistent and DRY.

Previously the code had a number of 6-line if-conditions and cascading if-elses that all check module versioning in a slightly different way. I've tried to identify the common logic and unify it all.

### Breaking Change Summary

Fix #7496 
Previous to this PR, having a module loaded with one version, meant that importing a different version of that module would give you the (wrong) loaded version. This change fixes the version check so that the loaded module is only used when it is the correct version. See https://github.com/PowerShell/PowerShell/issues/7496.

### Tests

Tests for this PR were already added in https://github.com/PowerShell/PowerShell/pull/7499. This PR just removes the `-Pending` flags from the tests around the breaking change (on the basis that the previous behaviour was a bug).

Fixes https://github.com/PowerShell/PowerShell/issues/7496.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/powershell/7125)

### CODE REVIEW

**Comment by iSazonov** [comment]
>Also, if our testing coverage on this should be improved, I'm happy to add tests to this PR 

To make a quality check of this change we are forced to do a lot of tests locally. So it makes sense to start by reviewing and extending the tests in the repo. I see that we don't actually have the tests for this code. I think we could add the new tests in new separate PR and then continue the PR. The `Import-Module.Tests.ps1` file is already large so we could create new `Import-Module-Versions.Tests.ps1`.

**Comment by rjmholt** [comment]
Ah so there *is* an Import-Module.Tests.ps1? It’s not under /test/powershell/engine/Module like I expected. Where is it?

Yes agreed we should bulk up the tests in another PR to merge before this one. I’ve been thinking about doing that but now I know there’s a test file I will see what we already have...

**Comment by iSazonov** [comment]
See  `test\powershell\Modules\Microsoft.PowerShell.Core\` subdirectory.

**Comment by anmenaga** [comment]
@rjmholt It would be good to see a green `[Feature]` test pass result. Can you please run it again?

**Comment by anmenaga** [comment]
@daxian-dbw @BrucePay We need to get some traction on this; this PR is sitting without a review for more than a month now...

**Comment by rjmholt** [comment]
I need to write tests for I would say. Will mark as WIP.

**Comment by iSazonov** [comment]
@rjmholt Please resolve the merge conflict.

**Comment by rjmholt** [comment]
I have now rebased this to include the 774 tests I wrote

**Comment by rjmholt** [comment]
I've removed the breaking change tag since I think another PR has already made the relevant change. This PR changes no tests and passes all existing.

**Comment by rjmholt** [comment]
Windows test failure is:
```
++++++?++++++
2018-10-10T23:48:47.4155281Z       [-] ResolveDestination for address 
2018-10-10T23:48:47.4163493Z         Expected strings to be the same, but they were different.
2018-10-10T23:48:47.4170377Z         String lengths are both 15.
2018-10-10T23:48:47.4182758Z         Strings differ at index 0.
2018-10-10T23:48:47.4190319Z         Expected: {factoryvm-az387}
2018-10-10T23:48:47.4198841Z         But was:  {aion.f2pool.com}
2018-10-10T23:48:47.4206302Z         -----------^
2018-10-10T23:48:47.4213664Z         171:             $result.Destination | Should -BeExactly $resolvedName
2018-10-10T23:48:47.4220802Z         at Invoke-Assertion, D:\a\1\s\src\powershell-win-core\bin\Release\netcoreapp2.1\win7-x64\publish\Modules\Pester\4.2.0\Functions\Assertions\Should.ps1: line 206
2018-10-10T23:48:47.4229034Z         at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Test-Connection.Tests.ps1: line 171
2018-10-10T23:48:52.2124211Z +++
```

Gotten it twice now. Building again.

**Comment by vexx32** [comment]
@rjmholt given that I am also seeing this issue in #7993, I think this is something that changed in the build outside your code.

Not sure exactly when this happened, though it appears to be a recent issue.

**Comment by rjmholt** [comment]
@vexx32 it looks like a CI infrastructure configuration problem -- some kind of DNS state has changed and the test depends on it

**Comment by iSazonov** [comment]
Fix for CI is in #7987

File: src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs [hunk 1]
```diff
@@ -401,21 +401,253 @@ internal List<PSModuleInfo> GetModules(ModuleSpecification[] fullyQualifiedName,
...
+        /// <returns>True if the module parameters match all given constraints, false otherwise.</returns>
+        internal static bool AreModuleFieldsMatchingConstraints(
+            string moduleName = null,
+            Guid? moduleGuid = null,
+            Version moduleVersion = null,
+            string requiredName = null,
+            Guid? requiredGuid = null,
+            Version requiredVersion = null,
+            Version minimumRequiredVersion = null,
+            Version maximumRequiredVersion = null)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs]: Does it really make sense to be so permissive?

**Review by daxian-dbw — APPROVED** [review]
LGTM except for the comment about the optional parameters.

**Comment by iSazonov** [comment]
I'd expect that we add new tests because of "breaking change".

**Comment by daxian-dbw** [comment]
@rjmholt Can you please summarize what the breaking change is in the PR description?

**Comment by rjmholt** [comment]
@iSazonov Tests were added in https://github.com/PowerShell/PowerShell/pull/7499. I added the tests for the new behaviour then and have removed the `-Pending` flag in this PR.

File: src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs (hunk 1 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs]: On the one hand, I don't think we need all the overloads I've included here currently, or at least not taking so many parameters.

On the other hand, the parameter set matches the problem domain:
- We have a `ModuleInfo` or its fields. In some cases we have no name, in others we have no guid, and we are never guaranteed that the version is not `null` by the caller
- We are given a set of module constraints (usually a `ModuleSpecification`) which we model as a hashtable with a number of fields that may or may not be there
- The method's contract is basically to take whatever parts of the module it knows about and see if they fit within whatever constraints it was given
- If there are no constraints, then anything matches

Does this make the behaviour permissive? Possibly. But I think it's well-reasoned and consistent, meaning we can depend on it in any future scenarios, rather than setting its behaviour to suit the current scenario and having to break it later.

**Review by daxian-dbw — APPROVED** [review]
Thanks for revisit some of the optional parameters.

**Comment by anmenaga** [comment]
@rjmholt Please address remaining couple of minor issues that Ilya pointed out and this will be ready.

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 2]
```diff
@@ -5059,34 +4988,19 @@ private bool ShouldModuleBeRemoved(PSModuleInfo module, string moduleNameInRemov
...
-            }
-            if (BaseGuid != null && !alreadyLoadedModule.Guid.Equals(this.BaseGuid))
-            {
-                // "alreadyLoadedModule" is a different module (than the one we want to load)
-                return false;
-            }
-
-            return true;
+            return ModuleIntrinsics.IsModuleMatchingConstraints(
+                alreadyLoadedModule,
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: Sorry, I wrote this some time ago but must have forgotten to press the "publish" button.

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 3]
```diff
@@ -5478,14 +5377,11 @@ internal ExternalScriptInfo GetScriptInfoForFile(string fileName, out string scr
                 // This will allow the search to continue and load a different version of the module.
                 if (Context.Modules.ModuleTable.TryGetValue(fileName, out module))
                 {
-                    if (BaseMinimumVersion != null && module.Version >= BaseMinimumVersion)
+                    if (!ModuleIntrinsics.IsVersionMatchingConstraints(module.Version, minimumVersion: BaseMinimumVersion, maximumVersion: BaseMaximumVersion))
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: See above comment

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 4]
```diff
@@ -5059,34 +4988,19 @@ private bool ShouldModuleBeRemoved(PSModuleInfo module, string moduleNameInRemov
             return false;
         }
 
+        /// <summary>
+        /// Checks if an already loaded module meets the constraints passed to the module cmdlet by the user.
+        /// </summary>
+        /// <param name="alreadyLoadedModule">The already loaded module that matched the name of the module to load.</param>
+        /// <returns>True if the pre-loaded module matches all GUID and version constraints provided, false otherwise.</returns>
         internal bool IsModuleAlreadyLoaded(PSModuleInfo alreadyLoadedModule)
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: This method has been in PowerShell with this name (and with this behaviour) since at least Win8. I would be happy to change it, so long as it's not going to cause any issues.

**Comment by anmenaga** [comment]
@adityapatwardhan this looks ready.

**Comment by iSazonov** [comment]
@rjmholt Please address two comments.

**Comment by adityapatwardhan** [comment]
@rjmholt Please address comments from @iSazonov. After that, is seems ready to merge.

**Comment by rjmholt** [comment]
@adityapatwardhan, @iSazonov Feedback addressed -- let me know if any further changes are required.

**Comment by adityapatwardhan** [comment]
@SteveL-MSFT  The newly added test is failing. It is marked as Feature and Feature tests were not run when the PR got merged.

**Comment by SteveL-MSFT** [comment]
Looking

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@rjmholt Please look CodeFactor issues and fix if they in your code.
